### PR TITLE
Promote IFormFile extension methods

### PR DESCRIPTION
**Pull request by khellang**
One alternative to fix half of #527 :sparkles:

### CODE REVIEW

File: src/Microsoft.AspNet.Http/Features/FormFile.cs [hunk 1]
```diff
@@ -2,12 +2,17 @@
...
+using System.Threading;
+using System.Threading.Tasks;
 using Microsoft.AspNet.Http.Internal;
 
 namespace Microsoft.AspNet.Http.Features.Internal
 {
     public class FormFile : IFormFile
     {
+        // Stream.CopyTo method uses 80KB as the default buffer size.
+        private const int DefaultBufferSize = 80*1024;
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Http/Features/FormFile.cs]: Spaces

File: src/Microsoft.AspNet.Http/Features/FormFile.cs [hunk 2]
```diff
@@ -20,29 +25,86 @@ public FormFile(Stream baseStream, long baseStreamOffset, long length, string na
...
+        }
+
+        /// <summary>
+        /// Asynchronously saves the contents of the uploaded file.
+        /// </summary>
+        /// <param name="path">The path of the file to create.</param>
+        /// <param name="cancellationToken"></param>
+        public async Task SaveAsAsync(string path, CancellationToken cancellationToken)
+        {
+            using (var fileStream = new FileStream(path, FileMode.Create))
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Http/Features/FormFile.cs]: Use File.Create instead

File: src/Microsoft.AspNet.Http/Features/FormFile.cs [hunk 3]
```diff
@@ -20,29 +25,86 @@ public FormFile(Stream baseStream, long baseStreamOffset, long length, string na
...
             return new ReferenceReadStream(_baseStream, _baseStreamOffset, Length);
         }
+
+        /// <summary>
+        /// Saves the contents of the uploaded file.
+        /// </summary>
+        /// <param name="path">The path of the file to create.</param>
+        public void SaveAs(string path)
+        {
+            using (var fileStream = new FileStream(path, FileMode.Create))
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Http/Features/FormFile.cs]: Use File.Create instead

**Comment by davidfowl** [comment]
I like it

File: src/Microsoft.AspNet.Http.Abstractions/IFormFile.cs [hunk 4]
```diff
@@ -10,18 +12,58 @@ namespace Microsoft.AspNet.Http
...
+        /// </summary>
+        /// <param name="path">The path of the file to create.</param>
+        Task SaveAsAsync(string path);
+
+        /// <summary>
+        /// Asynchronously saves the contents of the uploaded file.
+        /// </summary>
+        /// <param name="path">The path of the file to create.</param>
+        /// <param name="cancellationToken"></param>
+        Task SaveAsAsync(string path, CancellationToken cancellationToken);
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Http.Abstractions/IFormFile.cs]: The pattern we chose for cancellation token was to:

```C#
Task SaveAsAsync(string path, CancellationToken cancellationToken = default(CancellationToken)));
```

Comment (khellang) [inline src/Microsoft.AspNet.Http.Abstractions/IFormFile.cs]: Aigh. I thought optional params were frowned upon :frowning:

Comment (davidfowl) [inline src/Microsoft.AspNet.Http.Abstractions/IFormFile.cs]: They are generally but https://github.com/aspnet/Home/wiki/Engineering-guidelines#async-method-patterns

**Comment by davidfowl** [comment]
@khellang squash the commits and I'll merge it. Let's see what breaks!

**Comment by khellang** [comment]
Done :sparkles:

**Comment by khellang** [comment]
I guess this warrants an announcement?

File: src/Microsoft.AspNet.Http/Features/FormFile.cs [hunk 5]
```diff
@@ -20,29 +25,77 @@ public FormFile(Stream baseStream, long baseStreamOffset, long length, string na
...
+        }
+
+        /// <summary>
+        /// Asynchronously saves the contents of the uploaded file.
+        /// </summary>
+        /// <param name="path">The path of the file to create.</param>
+        /// <param name="cancellationToken"></param>
+        public async Task SaveAsAsync(string path, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            using (var fileStream = File.Create(path, DefaultBufferSize))
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Http/Features/FormFile.cs]: Almost done. You need to pass `FileOptions` and pass the async flag.

Comment (khellang) [inline src/Microsoft.AspNet.Http/Features/FormFile.cs]: :+1:

**Pull request merged**